File: src/06_PrintListInReversedOrder/06_PrintListInReversedOrder.py
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

import os
import sys
sys.path.append(os.path.split(sys.path[0])[0])

from Utilities import list_utils
from Utilities.list_utils import ListNode


class Solution:
    def reversePrint(self, head: ListNode, isRecur: bool = False) -> list[int]:
        if isRecur:
            return self.reversePrint_Recursively(head)
        else:
            return self.reversePrint_Iteratively(head) 

    # 使用递归
    def reversePrint_Recursively(self, head: ListNode):
        if head is None:
            return []
        return self.reversePrint_Recursively(head.next) + [head.val]

    # 使用基于栈的迭代
    def reversePrint_Iteratively(self, head: ListNode):
        stk = []
        while head is not None:
            stk.append(head.val)
            head = head.next

        # 逆序输出
        return stk[::-1]


class Test:
    def __init__(self):
        self.solution = Solution()

    def checkResult(self, head: ListNode):
        list_utils.printListNode(head)
        print(self.solution.reversePrint(head, True))
        print(self.solution.reversePrint(head, False))

    def Test1(self):
        print("\nTest1:\n")
        vals = [1,2,3,4,5]
        nodes = list_utils.createListNodesVec(vals)
        list_utils.connectListNodesVec(nodes)

        self.checkResult(nodes[0])

    # 只有一个结点的链表: 1
    def Test2(self):
        print("\nTest1:\n")
        node1 = list_utils.createListNode(1)

        self.checkResult(node1)

    # 空链表
    def Test3(self):
        print("\nTest1:\n")
        self.checkResult(None)

    def runTest(self):
        self.Test1()
        self.Test2()
        self.Test3()


if __name__ == "__main__":
    test = Test()
    test.runTest()
    

File: src/06_PrintListInReversedOrder/06_PrintListInReversedOrder.cpp
// 面试题6：从尾到头打印链表
// 题目：输入一个链表的头结点，从尾到头反过来打印出每个结点的值。
// 实例：
// 输入：head = [1,3,2]
// 输出：[2,3,1]

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(nullptr) {}
 * };
 */

#include <iostream>
#include <stack>
#include <vector>

#include "../Utilities/List.h"

using namespace std;

class Solution {
   public:
    vector<int> reversePrint( ListNode *pHead, bool isRecur = false ) {
        results.clear();

        if ( isRecur )
            reversePrint_Recursively( pHead );
        else
            reversePrint_Iteratively( pHead );

        return results;
    }

    // 使用递归
    void reversePrint_Recursively( ListNode *pHead ) {
        if ( pHead == nullptr ) return;
        if ( pHead->next != nullptr ) reversePrint_Recursively( pHead->next );
        results.push_back( pHead->val );
    }

    // 使用基于栈的迭代
    void reversePrint_Iteratively( ListNode *pHead ) {
        stack<int> stk;
        while ( pHead != nullptr ) {
            stk.push( pHead->val );
            pHead = pHead->next;
        }

        while ( !stk.empty() ) {
            results.push_back( stk.top() );
            stk.pop();
        }
    }

   private:
    vector<int> results;
};

class Test {
   public:
    void checkResult( ListNode *pHead ) {
        printListNode( pHead );
        vector<int> results;
        results = solution.reversePrint( pHead, true );
        printResults( results );
        results = solution.reversePrint( pHead, false );
        printResults( results );
    }

    // 1->2->3->4->5
    void Test1() {
        cout << "\nTest1 begins.\n";

        ValVec vals{1,2,3,4,5};
        ListNodeVec nodes = createListNodesVec(vals);
        connectListNodesVec(nodes);

        checkResult( nodes[0] );
        destroyList( nodes[0] );
    }

    // 只有一个结点的链表: 1
    void Test2() {
        cout << "\nTest2 begins.\n";

        ListNode *pNode1 = createListNode( 1 );
        checkResult( pNode1 );
        destroyList( pNode1 );
    }

    // 空链表
    void Test3() {
        cout << "\nTest3 begins.\n";
        checkResult( nullptr );
    }

    void runTest() {
        Test1();
        Test2();
        Test3();
    }

   private:
    Solution solution;

    template <typename T>
    void printResults( vector<T> &results ) {
        if ( results.empty() ) return;

        for ( T item : results )
            cout << item << " ";
        cout << endl;
    }
};

int main() {
    Test test;
    test.runTest();
    return 0;
}
